### PR TITLE
8346082: Output JVMTI agent information in hserr files

### PR DESCRIPTION
**Pull request by MBaesken**
We should output more information about the JVMTI agents in the hserr file.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8346082](https://bugs.openjdk.org/browse/JDK-8346082): Output JVMTI agent information in hserr files (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) Review applies to [fa422ce3](https://git.openjdk.org/jdk/pull/22706/files/fa422ce304fba1f7244d7c13aa4bb0f8861c55ad)
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) Review applies to [fa960c86](https://git.openjdk.org/jdk/pull/22706/files/fa960c86de21e02da5e3d571e13ee75002229ec9)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22706/head:pull/22706` \
`$ git checkout pull/22706`

Update a local copy of the PR: \
`$ git checkout pull/22706` \
`$ git pull https://git.openjdk.org/jdk.git pull/22706/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22706`

View PR using the GUI difftool: \
`$ git pr show -t 22706`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22706.diff">https://git.openjdk.org/jdk/pull/22706.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22706#issuecomment-2538391506)
</details>

### CODE REVIEW

File: src/hotspot/share/runtime/os.cpp [hunk 1]
```diff
@@ -1121,6 +1122,37 @@ void os::print_environment_variables(outputStream* st, const char** env_list) {
...
+      if (optionsinfo == nullptr) optionsinfo = "none";
+      if (pathinfo == nullptr) pathinfo = "none";
+      // jplis output too?
+      if (agent->is_initialized()) initinfo = "initialized";
+      st->print_cr("%s path:%s, %s, %s %s %s options:%s", agent->name(), pathinfo, loadinfo, initinfo, dyninfo, instrumentinfo, optionsinfo);
+    }
+  }
+#else
+  st->print_cr("JVMTI support not included in this JVM.");
+#endif
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/os.cpp:1153]: This is unreachable given the caller is already guarded. It is also unnecessary to report this.

File: src/hotspot/share/utilities/vmError.cpp [hunk 2]
```diff
@@ -1275,6 +1275,12 @@ void VMError::report(outputStream* st, bool _verbose) {
     os::print_dll_info(st);
     st->cr();
 
+#if INCLUDE_JVMTI
+  STEP_IF("printing jvmti agent infos", _verbose)
```
Comment (dholmes-ora) [inline src/hotspot/share/utilities/vmError.cpp:1279]:
```suggestion
  STEP_IF("printing jvmti agent info", _verbose)
```

**Review by dholmes-ora — APPROVED** [review]
Seems okay. Someone from serviceability should give their opinion on the content.

**Comment by larry-cable** [comment]
I do not understand the use case that this additional information solves for, nor does the bug provide any further enlightenment, I think this ER requires more rationale to motivate its inclusion.

File: src/hotspot/share/runtime/os.cpp [hunk 3]
```diff
@@ -1121,6 +1122,35 @@ void os::print_environment_variables(outputStream* st, const char** env_list) {
   }
 }
 
+void os::print_jvmti_agent_info(outputStream* st) {
+#if INCLUDE_JVMTI
+  // should return all kinds of JVMTI agents, but no xrun agents
+  const JvmtiAgentList::Iterator it =JvmtiAgentList::agents();
```
Comment (alexmenkov) [inline src/hotspot/share/runtime/os.cpp:1128]:
```suggestion
  const JvmtiAgentList::Iterator it = JvmtiAgentList::agents();
```

File: src/hotspot/share/runtime/os.cpp [hunk 4]
```diff
@@ -1121,6 +1122,35 @@ void os::print_environment_variables(outputStream* st, const char** env_list) {
...
+    if (agent != nullptr) {
+      if (first_agent) st->print_cr("JVMTI agents:");
+      first_agent = false;
+      const char* dyninfo = "";
+      const char* instrumentinfo = "";
+      const char* loadinfo = "not loaded";
+      const char* initinfo = "not initialized";
+      const char* optionsinfo = agent->options();
+      const char* pathinfo = agent->os_lib_path();
+      if (agent->is_dynamic()) dyninfo = "dynamic";
```
Comment (alexmenkov) [inline src/hotspot/share/runtime/os.cpp:1141]: maybe just initialize it as
```
      const char* dyninfo = agent->is_dynamic() ? "dynamic" : "";
```
(and the same for instrumentinfo/loadinfo/initinfo)

**Comment by MBaesken** [comment]
> I do not understand the use case that this additional information solves for, nor does the bug provide any further enlightenment, I think this ER requires more rationale to motivate its inclusion.

Currently the info about the JVMTI agents in the hserr/hsinfo is not good. You might see them in the so/dll list any maybe command-line output but this is far from ideal.  A little separate section gives more clarity.
Bad JVMTI agents (and even good ones) can influence what is going on at runtime so that info should be present.

File: src/hotspot/share/runtime/os.cpp [hunk 5]
```diff
@@ -1121,6 +1122,31 @@ void os::print_environment_variables(outputStream* st, const char** env_list) {
...
+      const char* dyninfo = agent->is_dynamic() ? "dynamic" : "";
+      const char* instrumentinfo = agent->is_instrument_lib() ? "instrumentlib" : "";
+      const char* loadinfo = agent->is_loaded() ? "loaded" : "not loaded";
+      const char* initinfo = agent->is_initialized() ? "initialized" : "not initialized";
+      const char* optionsinfo = agent->options();
+      const char* pathinfo = agent->os_lib_path();
+      if (optionsinfo == nullptr) optionsinfo = "none";
+      if (pathinfo == nullptr) pathinfo = "none";
+      // jplis output too?
+      st->print_cr("%s path:%s, %s, %s %s %s options:%s", agent->name(), pathinfo, loadinfo, initinfo, dyninfo, instrumentinfo, optionsinfo);
```
Comment (TheRealMDoerr) [inline src/hotspot/share/runtime/os.cpp:1144]: I think the spaces can be improved by removing them between the "%s"s and added them to the Strings above.

**Review by TheRealMDoerr — COMMENTED** [review]
Nice enhancement! Could you add an example output to the description, please?

File: src/hotspot/share/runtime/os.cpp [hunk 6]
```diff
@@ -1121,6 +1122,31 @@ void os::print_environment_variables(outputStream* st, const char** env_list) {
...
+  bool first_agent = true;
+  while (it.has_next()) {
+    const JvmtiAgent* agent = it.next();
+    if (agent != nullptr) {
+      if (first_agent) st->print_cr("JVMTI agents:");
+      first_agent = false;
+      const char* dyninfo = agent->is_dynamic() ? "dynamic" : "";
+      const char* instrumentinfo = agent->is_instrument_lib() ? "instrumentlib" : "";
+      const char* loadinfo = agent->is_loaded() ? "loaded" : "not loaded";
+      const char* initinfo = agent->is_initialized() ? "initialized" : "not initialized";
```
Comment (tstuefe) [inline src/hotspot/share/runtime/os.cpp:1138]: I would just print these as flags, its clearer and simpler to print. E.g.
```
"myagent path:  mylib.so dynamic:1 loaded:1 instrument_lib:0 initialized:1 xxx options xxx"
```

File: src/hotspot/share/runtime/os.cpp [hunk 7]
```diff
@@ -1121,6 +1122,31 @@ void os::print_environment_variables(outputStream* st, const char** env_list) {
...
 
+void os::print_jvmti_agent_info(outputStream* st) {
+#if INCLUDE_JVMTI
+  // should return all kinds of JVMTI agents, but no xrun agents
+  const JvmtiAgentList::Iterator it = JvmtiAgentList::agents();
+  bool first_agent = true;
+  while (it.has_next()) {
+    const JvmtiAgent* agent = it.next();
+    if (agent != nullptr) {
+      if (first_agent) st->print_cr("JVMTI agents:");
```
Comment (tstuefe) [inline src/hotspot/share/runtime/os.cpp:1133]: if no agents are loaded, I would print "JVMTI agents: none" unconditionally. Makes it more obvious than just a missing entry; that could be also an error.

**Review by tstuefe — APPROVED** [review]
Looks fine, two bikeshedding nits.

I approve now, up to you if you take my suggestions.

**Comment by MBaesken** [comment]
> Nice enhancement! Could you add an example output to the description, please?

Currently it looks like this

```
JVMTI agents:
asyncProfiler path:none, loaded, initialized   options:start,flat=10000,interval=50us,traces=1,event=wall,loglevel=none
```

(example is from a JVM started with async-profiler   :  `images/jdk/bin/java -agentlib:asyncProfiler=start,flat=10000,interval=50us,traces=1,event=wall,loglevel=none   ....`  )

**Comment by MBaesken** [comment]
>if no agents are loaded, I would print "JVMTI agents: none" unconditionally. Makes it more obvious than just a missing entry; >that could be also an error.

Maybe after the while loop something like this ?

```
  bool first_agent = true;
  while (it.has_next()) {
      ...
  }
  **if (first_agent) st->print_cr("JVMTI agents: none");**

```

**Comment by alexmenkov** [comment]
> > if no agents are loaded, I would print "JVMTI agents: none" unconditionally. Makes it more obvious than just a missing entry; >that could be also an error.
> 
> Maybe after the while loop something like this ?
> 
> ```
>   bool first_agent = true;
>   while (it.has_next()) {
>       ...
>   }
>   **if (first_agent) st->print_cr("JVMTI agents: none");**
> ```

You can do it before the loop:
```
if (it.has_next()) {
  st->print_cr("JVMTI agents:");
} else {
  st->print_cr("JVMTI agents: none");
}
```
and get rid of `first_agent` variable

**Comment by MBaesken** [comment]
When printing  'JVMTI agents: none' , should I maybe better iterate  `JvmtiAgentList::all() `  to get the xrun agents as well and be sure that there are  _really_ none ?
Unfortunately `JvmtiAgentList::all() `  is private , any objections making it public ?  Otherwise we probably need to call  2 iterations (agents() and xrun_agents()).

**Comment by TheRealMDoerr** [comment]
> again what's the use case?

Assume a customer uses a tool like the async profiler and causes a JVM crash by doing that. A supporter looks at the hs_err file and wonders about what might have happened. Such a hint would be very helpful.

**Comment by tstuefe** [comment]
> > jcmd to emit agent info
> 
> Okay, then let's avoid adding it to jcmd; for hserr/hsinfo adding the info is more important.
> 
> My idea was that currently we can do 'data dump' and 'load agents' with jcmd
> 
> ```
> JVMTI.agent_load
> JVMTI.data_dump
> ```
> 
> So for completeness, when you offer loading agents, it might be useful to be able to _**show**_ the current agents with the same tool jcmd. But on the other hand you are probably correct that this new option might not be used a lot so it is maybe not worth the effort.

Since we already have VM.info jcmd, which after your patch already shows jvmti info, a separate command would be redundant, no?

**Pull request closed without merging**